File: connector-typescript/src/testing.ts
import type {
  ConnectorContext,
  ConnectorHttpClient,
  ConnectorConfigReader,
  ConnectorLogger,
  HttpResponse,
} from './types.js';

interface MockHttpRule {
  status: number;
  body: unknown;
  headers?: Record<string, string>;
}

interface HttpCall {
  method: string;
  url: string;
  headers?: Record<string, string>;
  body?: unknown;
}

export interface TestHttpClient extends ConnectorHttpClient {
  /** All HTTP calls made during the test */
  calls: HttpCall[];
}

export interface CreateTestContextOptions {
  config?: Record<string, string>;
  httpMock?: Record<string, MockHttpRule>;
  agentId?: string;
  organizationId?: string;
}

function matchUrl(pattern: string, url: string): boolean {
  if (pattern === url) return true;
  if (pattern.endsWith('*')) {
    return url.startsWith(pattern.slice(0, -1));
  }
  return false;
}

/**
 * Create a mock ConnectorContext for testing connectors without running Council.
 *
 * @example
 * ```typescript
 * const ctx = createTestContext({
 *   config: { apiKey: 'test-key' },
 *   httpMock: {
 *     'https://api.example.com/*': { status: 200, body: { data: 'test' } },
 *   },
 * });
 *
 * const result = await myConnector.operations.fetch_data.handler({ query: 'test' }, ctx);
 * expect(result.data).toBe('test');
 * expect(ctx.http.calls).toHaveLength(1);
 * ```
 */
export function createTestContext(options: CreateTestContextOptions = {}): ConnectorContext & { http: TestHttpClient } {
  const { config = {}, httpMock = {}, agentId = 'test-agent', organizationId = 'test-org' } = options;
  const calls: HttpCall[] = [];

  async function mockRequest(method: string, url: string, opts?: { headers?: Record<string, string>; body?: unknown }): Promise<HttpResponse> {
    calls.push({ method, url, headers: opts?.headers, body: opts?.body });

    // Find matching mock rule
    for (const [pattern, rule] of Object.entries(httpMock)) {
      if (matchUrl(pattern, url)) {
        return {
          status: rule.status,
          data: rule.body,
          headers: rule.headers ?? {},
        };
      }
    }

    // No mock rule — return 404
    return { status: 404, data: { error: `No mock for ${method} ${url}` }, headers: {} };
  }

  const http: TestHttpClient = {
    calls,
    request: (method, url, opts) => mockRequest(method, url, opts),
    get: (url, opts) => mockRequest('GET', url, opts),
    post: (url, opts) => mockRequest('POST', url, opts),
    put: (url, opts) => mockRequest('PUT', url, opts),
    delete: (url, opts) => mockRequest('DELETE', url, opts),
  };

  const configReader: ConnectorConfigReader = {
    get: (key) => config[key],
    getRequired: (key) => {
      const val = config[key];
      if (val === undefined) throw new Error(`Missing required config: ${key}`);
      return val;
    },
    getAll: () => ({ ...config }),
  };

  const log: ConnectorLogger = {
    info: () => {},
    warn: () => {},
    error: () => {},
  };

  return { http, config: configReader, log, agentId, organizationId };
}

File: connector-python/council_connector_sdk/testing.py
"""Testing utilities for Council connectors."""

from __future__ import annotations

import fnmatch
from dataclasses import dataclass, field
from typing import Any

from .types import ConnectorContext, HttpResponse


@dataclass
class HttpCall:
    method: str
    url: str
    headers: dict[str, str] | None = None
    body: Any = None


@dataclass
class MockHttpRule:
    status: int
    body: Any
    headers: dict[str, str] | None = None


class TestHttpClient:
    def __init__(self, mocks: dict[str, MockHttpRule] | None = None):
        self.calls: list[HttpCall] = []
        self._mocks = mocks or {}

    def _find_mock(self, url: str) -> HttpResponse:
        for pattern, rule in self._mocks.items():
            if fnmatch.fnmatch(url, pattern):
                return HttpResponse(status=rule.status, data=rule.body, headers=rule.headers or {})
        return HttpResponse(status=404, data={"error": f"No mock for {url}"}, headers={})

    async def request(self, method: str, url: str, *, headers: dict[str, str] | None = None, body: Any = None, timeout: int | None = None) -> HttpResponse:
        self.calls.append(HttpCall(method=method, url=url, headers=headers, body=body))
        return self._find_mock(url)

    async def get(self, url: str, *, headers: dict[str, str] | None = None, timeout: int | None = None) -> HttpResponse:
        return await self.request("GET", url, headers=headers)

    async def post(self, url: str, *, headers: dict[str, str] | None = None, body: Any = None, timeout: int | None = None) -> HttpResponse:
        return await self.request("POST", url, headers=headers, body=body)

    async def put(self, url: str, *, headers: dict[str, str] | None = None, body: Any = None, timeout: int | None = None) -> HttpResponse:
        return await self.request("PUT", url, headers=headers, body=body)

    async def delete(self, url: str, *, headers: dict[str, str] | None = None, timeout: int | None = None) -> HttpResponse:
        return await self.request("DELETE", url, headers=headers)


class TestConfigReader:
    def __init__(self, config: dict[str, str] | None = None):
        self._config = config or {}

    def get(self, key: str) -> str | None:
        return self._config.get(key)

    def get_required(self, key: str) -> str:
        val = self._config.get(key)
        if val is None:
            raise KeyError(f"Missing required config: {key}")
        return val

    def get_all(self) -> dict[str, str]:
        return dict(self._config)


class TestLogger:
    def __init__(self):
        self.entries: list[dict[str, Any]] = []

    def info(self, message: str, data: dict[str, Any] | None = None) -> None:
        self.entries.append({"level": "info", "message": message, "data": data})

    def warn(self, message: str, data: dict[str, Any] | None = None) -> None:
        self.entries.append({"level": "warn", "message": message, "data": data})

    def error(self, message: str, data: dict[str, Any] | None = None) -> None:
        self.entries.append({"level": "error", "message": message, "data": data})


def create_test_context(
    *,
    config: dict[str, str] | None = None,
    http_mocks: dict[str, MockHttpRule] | None = None,
    agent_id: str = "test-agent",
    organization_id: str = "test-org",
) -> ConnectorContext:
    """Create a mock ConnectorContext for testing connectors without running Council.

    Usage::

        ctx = create_test_context(
            config={"api_key": "test-key"},
            http_mocks={
                "https://api.example.com/*": MockHttpRule(status=200, body={"data": "test"}),
            },
        )
        result = await my_connector.fetch_data(FetchParams(query="test"), ctx)
        assert result.data == "test"
        assert len(ctx.http.calls) == 1
    """
    return ConnectorContext(
        http=TestHttpClient(http_mocks),
        config=TestConfigReader(config),
        log=TestLogger(),
        agent_id=agent_id,
        organization_id=organization_id,
    )
